ProcessLib/BalanceProcess/BalanceProcessLocalAssembler.h

```diff
@@ -0,0 +1,154 @@
+          _bulk_face_id(bulk_face_ids[surface_element.getID()])
+    {
+        // only detJ of surface element is needed
+        using FemType =
+            NumLib::TemplateIsoparametric<ShapeFunction, ShapeMatricesType>;
+
+        FemType fe(*static_cast<const typename ShapeFunction::MeshElement*>(
+            &surface_element));
+
+        std::size_t const n_integration_points =
+            _integration_method.getNumberOfPoints();
+
+        std::vector<typename ShapeMatricesType::ShapeMatrices> shape_matrices;
+        shape_matrices.reserve(n_integration_points);
+        _detJ.reserve(n_integration_points);
+        for (std::size_t ip = 0; ip < n_integration_points; ++ip)
+        {
+            shape_matrices.emplace_back(ShapeFunction::DIM, GlobalDim,
+                                        ShapeFunction::NPOINTS);
+            // TODO computeShapeFunctions only detJ necessary
+            fe.computeShapeFunctions(
+                _integration_method.getWeightedPoint(ip).getCoords(),
+                shape_matrices[ip], GlobalDim);
+            _detJ.push_back(shape_matrices[ip].detJ);
+        }
+    }
+
+    /// Integration for the element with the id \c element_id.
+    /// @param element_id id of the element
+    /// @param x The global vector containing the values for numerical
+    /// integration.
+    /// @param balance PropertyVector the integration result will be stored
+    /// into, where balance has the same number of entries like mesh elements
+    /// exists.
+    /// @param bulk_process Reference to the bulk process is used to compute the
+    /// flux.
+    void integrate(std::size_t element_id,
+                   GlobalVector const& x,
+                   MeshLib::PropertyVector<double>& balance,
+                   Process const& bulk_process) override
+    {
+        SpatialPosition pos;
+        pos.setElementID(element_id);
+
+        auto surface_element_normal =
+            MeshLib::FaceRule::getSurfaceNormal(&_surface_element);
+        surface_element_normal.normalize();
+
+        std::size_t const n_integration_points =
+            _integration_method.getNumberOfPoints();
+        // balance[id_of_element] =
+        //   int_{\Gamma_e} \alpha \cdot grad variable \cdot normal \d \Gamma
+        for (std::size_t ip(0); ip < n_integration_points; ip++)
+        {
+            pos.setIntegrationPoint(ip);
+            auto const& wp = _integration_method.getWeightedPoint(ip);
+
+            auto const bulk_element_point = getBulkElementPoint(
+                bulk_process.getMesh(), _bulk_element_id, _bulk_face_id, wp);
+            auto const bulk_flux =
+                bulk_process.getFlux(_bulk_element_id, bulk_element_point, x);
+
+            for (std::size_t component_id(0);
+                 component_id < balance.getNumberOfComponents();
+                 ++component_id)
+            {
+                // TODO find solution for 2d case
```

Comment:
The element normal is calculated in the physical space, not in natural coordinates. If the element is not in the x-y-plane then I am not convinced that the z-component of the element normal vanishs.